File: src/Logikschicht/TemperatureController.cpp
#include "TemperatureController.h"
#include <iostream>

TemperatureController::TemperatureController(SensorInterface* s, HeaterControl* h, SafetyManager* sa, DisplayDriver* d)
    : sensor(s), heater(h), safety(sa), display(d), sollTemperatur(0.0), istTemperatur(0.0) {}

void TemperatureController::SollIstVergleich() {
    istTemperatur = sensor->leseTemperatur();
    display->aktualisiereAnzeige(istTemperatur);

    // Überprüft auf Sicherheitsgrenzen
    if (safety->Uebertemperaturschutz(istTemperatur)) {
        heater->steuereHeizelement(0); // Heizung ausschalten
        safety->Sicherheitsabschaltung();
    }
}

void TemperatureController::RegelungHeizleistung() {
    double differenz = sollTemperatur - istTemperatur;

    // Einfache Regelung (Proportional)
    double leistung = differenz * 10; // P-Regler, Faktor 10 = Verstärkung

    if (leistung < 0) leistung = 0;
    if (leistung > 100) leistung = 100; // Maximalleistung

    heater->steuereHeizelement(leistung);
}

File: src/Code Python Sprint 1/Logikschicht/temperature_controller.py
class TemperatureController:
    def __init__(self, sensor, heater, safety, display):
        self.sensor = sensor
        self.heater = heater
        self.safety = safety
        self.display = display
        self.target_temp = 0

    def set_target_temperature(self, temp):
        self.target_temp = temp

    def soll_ist_vergleich(self):
        self.current_temp = self.sensor.read_temperature()

    def regelung_heizleistung(self):
        if self.current_temp < self.target_temp:
            self.heater.on()
        else:
            self.heater.off()
